text/0000-projections-lifetimes-and-wf.md

```diff
@@ -0,0 +1,980 @@
+implementation can be both unsound ([#24622]), inconvenient
+([#23442]), and surprising ([#21748], [#25692]). The changes are as follows:
+
+- Simplify the outlives relation to be syntactically based.
+- Specify improved rules for the outlives relation and projections.
+- Specify more specifically where WF bounds are enforced, covering
+  several cases missing from the implementation.
+
+The proposed changes here have been tested and found to cause only a
+modest number of regressions (about two dozen root regressions were
+previously found on crates.io; however, that run did not yet include
+all the provisions from this RFC; updated numbers coming soon). In
+order to minimize the impact on users, the plan is to first introduce
+the changes in two stages:
+
+1. Initially, warnings will be issued for cases that violate the rules
+   specified in this RFC. These warnings are not lints and cannot be
+   silenced except by correcting the code such that it type-checks
+   under the new rules.
+2. After one release cycle, those warnings will become errors.
+
+Note that although the changes do cause regressions, they also cause
+some code (like that in [#23442]) which currently gets errors to
+compile successfully.
+
+# Motivation
+
+### TL;DR
+
+This is a long detailed RFC that is attempting to specify in some
+detail aspects of the type system that were underspecified or buggily
+implemented before. This section just summarizes the effect on
+existing Rust code in terms of changes that may be required.
+
+**Warnings first, errors later.** Although the changes described in
+this RFC are necessary for soundness (and many of them are straight-up
+bugfixes), there is some impact on existing code. Therefore the plan
+is to first issue warnings for a release cycle and then transition to
+hard errors, so as to ease the migration.
+
+**Associated type projections and lifetimes work more smoothly.** The
+current rules for relating associated type projections (like `T::Foo`)
+and lifetimes are somewhat cumbersome. The newer rules are more
+flexible, so that e.g. we can deduce that `T::Foo: 'a` if `T: 'a`, and
+similarly that `T::Foo` is well-formed if `T` is well-formed. As a
+bonus, the new rules are also sound. ;)
+
+**The outlives relation is simpler.** The older definition for the
+outlives relation `T: 'a` was rather subtle. The new rule basically
+says that if all type/lifetime parameters appearing in the type `T`
+must outlive `'a`, then `T: 'a` (though there can also be other ways
+for us to decide that `T: 'a` is valid, such as in-scope where
+clauses). So for example `fn(&'x X): 'a` if `'x: 'a` and `X: 'a`
+(presuming that `X` is a type parameter). The older rules were based
+on what kind of data was actually *reachable*, and hence accepted this
+type (since no data of `&'x X` is reachable from a function pointer).
+This change primarily affects struct declarations, since they may now
+require additional outlives bounds:
```

Comment:
Why is it a good idea to do that? It seems kind of weird to require the argument type of a function to outlive `'a`, just for the function pointer itself to do so.

I also have the impression that the example below is not about outliving, but about well-formedness. Of course `fn(&'a X)` should only be well-formed if `X: 'a`, but how is that related to the entire type outliving `'a`?